Introduced UI to list domain suggestions.

usamahamid opened the pull request:
This is one of the screens required for [custom domain association](https://github.com/wordpress-mobile/WordPress-Android/issues/7923)

This PR introduces a new screen, in which User can be search and select one of the available domain names (identified from `/v1.1/domains/suggestions/` API). The navigation to this UI is not included yet, this is intended to keep this PR from affecting any working implementation.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/viewmodel/domainregister/DomainSuggestionsViewModel.kt:
```diff
@@ -0,0 +1,130 @@
+
+    private fun fetchSuggestions() {
+        // Disable Load more
+        suggestions = ListState.Loading(suggestions, false)
```
Do we need to explicitly set `canLoadMore` attribute to false? False is a default value, isn't it?

malinajirka commented on WordPress/src/main/java/org/wordpress/android/viewmodel/domainregister/DomainSuggestionsViewModel.kt:
```diff
@@ -0,0 +1,130 @@
+        // Disable Load more
+        suggestions = ListState.Loading(suggestions, false)
+
+        val suggestDomainsPayload = SuggestDomainsPayload(searchQuery, false, false, true, 30)
```
We might consider creating a constant for the `30`. Wdyt?

malinajirka commented:
Thank you @usamahamid! Good job;). 

I'm done with my first pass. I haven't tested the app though, as it's not tied to the UI yet. I've left few comments, most of which are just my preference.

Are you planning to create another PR with unit tests?

usamahamid commented on WordPress/src/main/java/org/wordpress/android/ui/domainregister/suggestionslist/DomainSuggestionsAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+
+    private fun onDomainSuggestionSelected(suggestion: DomainSuggestionResponse?, position: Int) {
+        selectedPosition = position
+        notifyDataSetChanged()
```
Of course. 
We just have to update `selectedPosition` and `previouslySelectedPosition`. I have updated the same in 1c00bdb.

usamahamid commented on WordPress/src/main/java/org/wordpress/android/ui/domainregister/suggestionslist/DomainSuggestionsAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+    internal fun updateSuggestionsList(items: List<DomainSuggestionResponse>) {
+        list.clear()
+        list.addAll(items)
+        notifyDataSetChanged()
```
Since we are resetting the whole list, it might be easier and cheaper (performance wise) to call `notifyDataSetChanged()`. 
`notifyItemRangeChanged` will work as well, but the size of the new suggestions list might be different, which might cause problems. What do you think?

malinajirka replied:
Fair enough;)!

usamahamid commented on WordPress/src/main/java/org/wordpress/android/ui/domainregister/suggestionslist/DomainSuggestionsFragment.kt:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        outState.putSerializable(WordPress.SITE, viewModel.site)
```
Of course, since we are not updating in `SiteModel` here. It need not be preserved in the `instanceState`.

usamahamid commented on WordPress/src/main/java/org/wordpress/android/ui/domainregister/suggestionslist/DomainSuggestionsFragment.kt:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    private fun setupObservers() {
+        viewModel.suggestionsLiveData.observe(this, Observer {
```
Understandable, I have updated the same in 56af118.

usamahamid commented:
@malinajirka, I have updated most of PR as per your comments.
Please do check it out, when have a free moment.

I have included a new row in my site page to navigate into domain suggestions listing (for testing purposes). 
I will be raising another PR with the Mockito test cases.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/domainregister/suggestionslist/DomainSuggestionsFragment.kt:
```diff
@@ -0,0 +1,84 @@
+            ToastUtils.showToast(activity, "Still under development.")
+        }
+        domainSearchEditText.setOnEditorActionListener { view, actionId, _ ->
+            if (actionId == EditorInfo.IME_ACTION_DONE) {
```
I've missed this reading the code, but testing the functionality on an emulator revealed it. 

We might want to search as the user types not just when they press `Done`. That's why we have a handler in the ViewModel, right (the handler is used in submitSearch(..))? Also we use this approach in other parts of the app, so we should keep it consistent (eg. My Site -> Media -> Search)

usamahamid commented:
@malinajirka, I have updated the PR to include a `TextWatcher` as you mentioned (which is a definite improvement). :+1: 
I have replied to one of your comments, Please check it out and tell me what you think. :slightly_smiling_face:

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/domainregister/suggestionslist/DomainSuggestionsFragment.kt:
```diff
@@ -0,0 +1,84 @@
+            suggestionsListProgress.visibility = if (it == true) View.VISIBLE else View.GONE
+        })
+        viewModel.suggestionsLiveData.observe(this, Observer {
+            if (it != null) {
```
Thank you @usamahamid for bringing this up! Definitely an interesting idea! I wouldn't delay this PR on such a minor issue. We can introduce the `NonNullObserver` anytime later if we see same null check pattern reoccurring.

malinajirka commented:
Thank you for fixing the issues @usamahamid!

I'm sorry, but I have overlooked one issue, which we should fix before we merge this PR.

I've been reviewing the code and functionality, but I totally forgot to check the design. I believe some of the colors don't match the design
- background, radio buttons and the `choose domain` button colors don't match

There are two more tiny issues, which are not in the proposed design, but we should probably try to be consistent with the SiteCreation flow.
- the color of the magnifier is a bit lighter comparing to the SiteCreation flow
- the progress indicator is in the center of the screen (checkout the indicator in the SiteCreation https://gyazo.com/2d0910e1608b6748e84bb15cef7e5d73)

usamahamid commented:
@malinajirka, I have updated the PR with the UI changes you requested (Seems, I missed to update the theme for the new activity). 
Additionally I have fixed a minor bug where the search query is reset if it is empty, during orientation change.

The PR is ready for another pass. Please do check it out, when you have free moment.

usamahamid commented:
@malinajirka, I have updated the layout design to match the Site Creation layout and it is ready for another pass.
Please do check it out, when you have a free moment. :+1:

malinajirka commented:
Thank you @usamahamid! I've just done a final test and it looks awesome:)!
